ccb_vaultic/ccb_vaultic.c

```diff
@@ -0,0 +1,1021 @@
+        /*
+         *  info->hash.flag | WC_HASH_FLAGS_WILL_COPY --> Buffer entire message
+         *  info->hash.in != NULL                     --> Update
+         *  info->hash.digest != NULL                 --> Final
+         */
+        {
+            /* New/different hashtype than last time? */
+            if (c->hash_type != info->hash.type) {
+                /* New/different hash than last time.  Erase state */
+#if defined(CCBVAULTIC_DEBUG_ALL)
+                printf("   New Hash\n");
+#endif
+                if (c->m)
+                    free(c->m);
+                c->m         = NULL;
+                c->m_len     = 0;
+                c->hash_type = info->hash.type;
+            }
+
+            /* Update needed? */
+            /* III Buffering ALL messages, not just WILL_COPY */
+            if (    (info->hash.in != NULL) &&
+                    (info->hash.inSz > 0)) {
+                /* Buffer data */
+                if (c->m) {
+                    /* Realloc and add new data in */
+                    void *new_buf = realloc(c->m, c->m_len + info->hash.inSz);
+                    if (new_buf == NULL) {
+                        /* Failure to allocate.  Must return error */
+#if defined(CCBVAULTIC_DEBUG_ALL)
+                        printf("   Failed to realloc. New size:%lu\n",
+                                c->m_len + info->hash.inSz);
+#endif
+                        rc = MEMORY_E;
+                        break;
+                    }
+                    c->m = new_buf;
+#if defined(CCBVAULTIC_DEBUG_ALL)
+                    printf("   Realloc to %p. New size:%lu\n",
+                            c->m, c->m_len + info->hash.inSz);
+#endif
+                    } else {
+                    c->m = malloc(info->hash.inSz);
+                    if (!c->m) {
+                        /* Failure to allocate.  Must return error */
+#if defined(CCBVAULTIC_DEBUG_ALL)
+                        printf("   Failed to alloc. Size:%u\n",
+                                info->hash.inSz);
+#endif
+                        rc = MEMORY_E;
+                        break;
+                    }
+#if defined(CCBVAULTIC_DEBUG_ALL)
+                    printf("   Alloc to %p. Size:%u\n", c->m, info->hash.inSz);
+#endif
+                    c->m_len = 0;
+                }
+                memcpy(c->m + c->m_len, info->hash.in, info->hash.inSz);
+                c->m_len += info->hash.inSz;
+#if defined(CCBVAULTIC_DEBUG_ALL)
+                printf("   Buffered to %p. Buffer size:%lu\n", c->m, c->m_len);
+#endif
+                rc = 0;
+            }
+
+            /* Finalize needed? */
+            if (info->hash.digest) {
+                /* Initialize for Hashing */
+                VLT_U8 sha_out_len=0;
+                VLT_ALGO_PARAMS sha256_algo_params = {0};
+                sha256_algo_params.u8AlgoID = VLT_ALG_DIG_SHA256;
+#ifdef CCBVAULTIC_DEBUG_TIMING
+                ts[0]=now();
+#endif
+                int vlt_rc=0;
```

Comment:
NIT: Portability. Variables declared inline. Not following coding standards for spaces and braces. Do you want to clean this up or request an exception?

Reply:
I'll fix all of this.  The error handling was especially weak here and I simply added to help debug what was going on.  I'll fix the error handling and make it conformant.